File: iphone-code/code/Stasher/STLogInViewController.m
//
//  STLogInViewController.m
//  Stasher
//
//  Created by bhushan on 13/11/14.
//  Copyright (c) 2014 OAB. All rights reserved.
//

#import "STLogInViewController.h"
#import "STBaseTabBarController.h"

@interface STLogInViewController ()

@end

@implementation STLogInViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self.headerLabel setFont:[UIFont FontGothamRoundedMediumWithSize:16.0f]];
    [self.txtFieldPassword setFont:[UIFont FontForTextFields]];
    [self.txtFieldPassword setTextColor:[UIColor stasherTextColor]];
    [self.txtFieldUsernameId setFont:[UIFont FontForTextFields]];
    [self.txtFieldUsernameId setTextColor:[UIColor stasherTextColor]];
    [self.txtFieldPassword setAutocorrectionType:UITextAutocorrectionTypeNo];
    [self.txtFieldUsernameId setAutocorrectionType:UITextAutocorrectionTypeNo];
    self.txtFieldPassword.attributedPlaceholder = [[NSAttributedString alloc] initWithString:self.txtFieldPassword.placeholder
                                                                      attributes:@{
                                                                                   NSForegroundColorAttributeName:[UIColor stasherTextFieldPlaceHolderColor],
                                                                                   NSFontAttributeName : [UIFont FontForTextFields]
                                                                                   }];
    self.txtFieldUsernameId.attributedPlaceholder = [[NSAttributedString alloc] initWithString:self.txtFieldUsernameId.placeholder
                                                                                  attributes:@{
                                                                                               NSForegroundColorAttributeName:[UIColor stasherTextFieldPlaceHolderColor],
                                                                                               NSFontAttributeName : [UIFont FontForTextFields]
                                                                                               }];
    
    [self.logInButton.titleLabel setFont:[UIFont FontForButtonsWithSize:11.0]];
    [self.forgotPasswordButton.titleLabel setFont:[UIFont FontForButtonsWithSize:11.0]];
    [self.forgotPasswordButton.titleLabel setTextColor:[UIColor stasherTextColor]];
    if (IS_STANDARD_IPHONE_6) {
        [self.logInViaFacebooButtonk.titleLabel setFont:[UIFont FontForButtonsWithSize:8.0f]];
    }else{
        [self.logInViaFacebooButtonk.titleLabel setFont:[UIFont FontForButtonsWithSize:10.0f]];
    }
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:YES];
    [[STSharedCustoms sharedAddGestureInstanceWithDelegate:self] addRightSwipeGestureRecognizerToMe];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

#pragma mark ----- Actions

- (IBAction)logInButtonPressed:(id)sender
{
    [self.view setUserInteractionEnabled:NO];
    [self.txtFieldPassword resignFirstResponder];
    [self performSelector:@selector(logIn) withObject:self afterDelay:0.05];
}

- (void)logIn
{
    if ([self.txtFieldPassword.text validateNotEmpty] && [self.txtFieldUsernameId.text validateNotEmpty]) {
        if ([self.txtFieldUsernameId.text validateEmail] || [self.txtFieldUsernameId.text validateNotEmpty]) {
            httpReq = [[BOABHttpReq alloc] initBoabReqWithDelegate:self shouldShowActivityIndicatorView:YES];
            NSMutableDictionary *paramsDict = [[NSMutableDictionary alloc] init];
            [paramsDict setObject:self.txtFieldUsernameId.text forKey:kParamKeyUsername];
            [paramsDict setObject:self.txtFieldPassword.text forKey:kParamKeyPassword];
            [paramsDict setObject:kAPIActionLogin forKey:kParamKeyAction];
            [httpReq sendAsyncPostRequest:STASHER_BASE_URL params:paramsDict json:YES];
        }
        else{
            [UIAlertView showWithTitle:@""
                               message:@"Enter a valid E-mail address."
                     cancelButtonTitle:@"OK"
                     otherButtonTitles:nil
                              tapBlock:^(UIAlertView *alertView, NSInteger buttonIndex) {
                                  if (buttonIndex == [alertView cancelButtonIndex]) {
                                  }
                              }];
        }
    }
    else{
        [UIAlertView showWithTitle:@""
                           message:@"All fields are mandatory."
                 cancelButtonTitle:@"OK"
                 otherButtonTitles:nil
                          tapBlock:^(UIAlertView *alertView, NSInteger buttonIndex) {
                              if (buttonIndex == [alertView cancelButtonIndex]) {
                              }
                          }];
    }
}
- (IBAction)forgotPasswordButtonPressed:(id)sender
{
    forgotPasswordVC = [[STForgotPasswordViewController alloc] initWithNibName:@"STForgotPasswordViewController" bundle:[NSBundle mainBundle]];
    forgotPasswordVC.delegate = self;
    [forgotPasswordVC.view setFrame:self.view.frame];
    [self.view addSubview:forgotPasswordVC.view withAnimation:YES];
}

- (IBAction)LogInViaFacebookButtonPressed:(id)sender
{
    [[STLogInManager sharedInstance] setDelegate:self];
    [[STLogInManager sharedInstance] openSessionWithAllowLoginUI:YES];
}

- (void) sendPaswordToEmailButtonPressed
{
    if (forgotPasswordVC != nil) {
        [forgotPasswordVC.view removeFromSuperviewwithAnimation:YES];
    }
}

- (void) forgetpasswordCanceled
{
    if (forgotPasswordVC != nil) {
        [forgotPasswordVC.view removeFromSuperviewwithAnimation:YES];
    }
}

- (IBAction)backButtonPressed:(id)sender
{
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)rightGestureHandle
{
    [self.navigationController popViewControllerAnimated:YES];
    [STSharedCustoms sharedAddGestureInstanceWithDelegate:nil];
}

#pragma mark ----- Custom Methods
- (void) launchUserBaseTabsScreen
{
    UIStoryboard* storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:[NSBundle mainBundle]];
    STBaseTabBarController *baseTabBarController = [storyboard instantiateViewControllerWithIdentifier:@"STBaseTabBarController"];
    [self.navigationController pushViewController:baseTabBarController animated:YES];
}

#pragma mark ----- Textfield Methods
- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    if (textField == self.txtFieldUsernameId) {
        [self.txtFieldPassword becomeFirstResponder];
    }
    else
        [textField resignFirstResponder];
    return YES;
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    BOOL shouldChange = TRUE;
    if (textField == self.txtFieldPassword) {
        if ([string validateAlphanumeric]) {
            shouldChange = TRUE;
        }
        else{
            shouldChange = FALSE;
        }
    }
    
    return shouldChange;
}

#pragma mark ----- LogInManagerDelegate

- (void)facebookLogInSuccessfulWithUserDictionary:(NSDictionary *)userInfoDict
{
    NSLog(@"facebooklogIn Dict %@",userInfoDict);
    
    if (userInfoDict != nil) {
        NSMutableDictionary* localUserInfoDict = [[NSMutableDictionary alloc] init];
        [localUserInfoDict setObject:[userInfoDict objectForKey:@"first_name"] forKey:kParamKeyFirstname];
        [localUserInfoDict setObject:[userInfoDict objectForKey:@"last_name"] forKey:kParamKeyLastname];
        [localUserInfoDict setObject:[userInfoDict objectForKey:@"gender"] forKey:kParamKeyGender];
        [localUserInfoDict setObject:[userInfoDict objectForKey:@"email"] forKey:kParamKeyEmail];
        //[localUserInfoDict setObject:[[userInfoDict objectForKey:@"location"] objectForKey:@"name"] forKey:kParamKeyCountry];
        NSDateFormatter *dateFormat = [[NSDateFormatter alloc] init];
        [dateFormat setDateFormat:@"MM/dd/yyyy"];
        NSDate *theDate = [dateFormat dateFromString:[userInfoDict objectForKey:@"birthday"]];
        NSDateFormatter *dateFormat2 = [[NSDateFormatter alloc] init];
        [dateFormat2 setDateFormat:@"yyyy-MM-dd"];
        if ([dateFormat2 stringFromDate:theDate])
            [localUserInfoDict setObject:[dateFormat2 stringFromDate:theDate] forKey:kParamKeyDateOfBirth];
        if ([[NSData dataWithContentsOfURL:[NSURL URLWithString:[NSString stringWithFormat:@"http://graph.facebook.com/%@/picture",[userInfoDict objectForKey:@"id"]]]] base64EncodedStringWithOptions:0] != nil) {
            [localUserInfoDict setObject:[[NSData dataWithContentsOfURL:[NSURL URLWithString:[NSString stringWithFormat:@"http://graph.facebook.com/%@/picture",[userInfoDict objectForKey:@"id"]]]] base64EncodedStringWithOptions:0]  forKey:kParamKeyAvatar];
        }
        if (httpReq) {
            httpReq.delegate = nil;
            httpReq = nil;
        }
        httpReq = [[BOABHttpReq alloc] initBoabReqWithDelegate:self shouldShowActivityIndicatorView:YES];
        [localUserInfoDict setObject:kAPIActionFacebookLogIn forKey:kParamKeyAction];
        [httpReq sendAsyncPostRequest:STASHER_BASE_URL params:localUserInfoDict json:YES];
    }
}


#pragma mark - BOABHttpReqDelegates

- (void)BOABHttpReqFinishedWithResponse:(NSData*)resonseData andConnection:(BOABURLConnection*)conn
{
    [self.view setUserInteractionEnabled:YES];
     NSString *actionName = [[conn.userInfo objectForKey:KParamsKey] objectForKey:kParamKeyAction];
    if ([actionName isEqualToString:kAPIActionFacebookLogIn]) {
        [[NSUserDefaults standardUserDefaults] setObject:@"YES" forKey:@"IsFaceBookLogIn"];
    }else{
        [[NSUserDefaults standardUserDefaults] setObject:@"NO" forKey:@"IsFaceBookLogIn"];
    }
    [[NSUserDefaults standardUserDefaults] synchronize];
    
    if (resonseData) {
        NSDictionary *responseDictionary = [NSJSONSerialization JSONObjectWithData:resonseData options:kNilOptions error:nil];
        if (responseDictionary) {
            if ([responseDictionary objectForKey:@"usedetails"] && [[responseDictionary objectForKey:@"usedetails"] isKindOfClass:[NSDictionary class]]) {
                [[STUserIdentity sharedInstance] setUserInformationDictionary:[[responseDictionary objectForKey:@"usedetails"] mutableCopy]];
                NSMutableDictionary *logInInfoDictionary = [[responseDictionary objectForKey:@"usedetails"] mutableCopy];
                [logInInfoDictionary setObject:[NSNumber numberWithInt:1] forKey:kUserDefaultsIsUserLoggedIn];
                [[STLogInManager sharedInstance] updateUserDefaultsInLogInManagerWithDictionary:logInInfoDictionary];
                [self launchUserBaseTabsScreen];
            }
            else if ([responseDictionary objectForKey:@"error"]){
                [UIAlertView showWithTitle:@""
                                   message:[[responseDictionary objectForKey:@"error"] objectForKey:@"message"]
                         cancelButtonTitle:@"OK"
                         otherButtonTitles:nil
                                  tapBlock:^(UIAlertView *alertView, NSInteger buttonIndex) {
                                      if (buttonIndex == [alertView cancelButtonIndex]) {
                                      }
                                  }];
            }
        }
    }
}

- (void)BOABHttpReqFailedWithError:(NSError*)error
{
    [self.view setUserInteractionEnabled:YES];
    if (error) {
        
    }
}


@end

File: swift-code/Stasher/STLogInViewController.swift
//  Converted with Swiftify v1.0.6292 - https://objectivec2swift.com/
//
//  STLogInViewController.swift
//  Stasher
//
//  Created by bhushan on 13/11/14.
//  Copyright (c) 2014 OAB. All rights reserved.
//
import UIKit
class STLogInViewController: UIViewController, ForgotPasswordDelegate, STSharedCustomsDelegate, LogInManagerDelegate {
    var httpReq: BOABHttpReq?
    var forgotPasswordVC: STForgotPasswordViewController?

    @IBOutlet var logInButton: UIButton!
    @IBOutlet var forgotPasswordButton: UIButton!
    @IBOutlet var txtFieldUsernameId: UITextField!
    @IBOutlet var txtFieldPassword: UITextField!
    @IBOutlet var headerLabel: UILabel!
    @IBOutlet var logInViaFacebooButtonk: UIButton!


    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        headerLabel.font = UIFont.fontGothamRoundedMedium(withSize: 16.0)
        txtFieldPassword.font = UIFont()
        txtFieldPassword.textColor = UIColor.stasherText()
        txtFieldUsernameId.font = UIFont()
        txtFieldUsernameId.textColor = UIColor.stasherText()
        txtFieldPassword.autocorrectionType = .no
        txtFieldUsernameId.autocorrectionType = .no
        txtFieldPassword.attributedPlaceholder = NSAttributedString(string: txtFieldPassword.placeholder, attributes: [NSForegroundColorAttributeName: UIColor.stasherTextFieldPlaceHolder(), NSFontAttributeName: UIFont()])
        txtFieldUsernameId.attributedPlaceholder = NSAttributedString(string: txtFieldUsernameId.placeholder, attributes: [NSForegroundColorAttributeName: UIColor.stasherTextFieldPlaceHolder(), NSFontAttributeName: UIFont()])
        logInButton.titleLabel?.font = UIFont(size: 11.0)
        forgotPasswordButton.titleLabel?.font = UIFont(size: 11.0)
        forgotPasswordButton.titleLabel?.textColor = UIColor.stasherText()
        if IS_STANDARD_IPHONE_6 {
            logInViaFacebooButtonk.titleLabel?.font = UIFont(size: 8.0)
        }
        else {
            logInViaFacebooButtonk.titleLabel?.font = UIFont(size: 10.0)
        }
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(true)
        STSharedCustoms.sharedAddGestureInstance(withDelegate: self).addRightSwipeGestureRecognizerToMe()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    /*
    #pragma mark - Navigation
    
    // In a storyboard-based application, you will often want to do a little preparation before navigation
    - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
        // Get the new view controller using [segue destinationViewController].
        // Pass the selected object to the new view controller.
    }
    */
// MARK: ----- Actions

    @IBAction func log(inButtonPressed sender: Any) {
        view.userInteractionEnabled = false
        txtFieldPassword.resignFirstResponder()
        performSelector(#selector(self.logIn), withObject: self, afterDelay: 0.05)
    }

    func logIn() {
        if txtFieldPassword.text.validateNotEmpty() && txtFieldUsernameId.text.validateNotEmpty() {
            if txtFieldUsernameId.text.validateEmail() || txtFieldUsernameId.text.validateNotEmpty() {
                httpReq = BOABHttpReq.initBoabReq(withDelegate: self, shouldShowActivityIndicatorView: true)
                var paramsDict = [AnyHashable: Any]()
                paramsDict[kParamKeyUsername] = txtFieldUsernameId.text
                paramsDict[kParamKeyPassword] = txtFieldPassword.text
                paramsDict[kParamKeyAction] = kAPIActionLogin
                httpReq?.sendAsyncPostRequest(STASHER_BASE_URL, params: paramsDict, json: true)
            }
            else {
                UIAlertView.show(withTitle: "", message: "Enter a valid E-mail address.", cancelButtonTitle: "OK", otherButtonTitles: nil, tapBlock: {(_ alertView: UIAlertView, _ buttonIndex: Int) -> Void in
                    if buttonIndex == alertView.cancelButtonIndex {

                    }
                })
            }
        }
        else {
            UIAlertView.show(withTitle: "", message: "All fields are mandatory.", cancelButtonTitle: "OK", otherButtonTitles: nil, tapBlock: {(_ alertView: UIAlertView, _ buttonIndex: Int) -> Void in
                if buttonIndex == alertView.cancelButtonIndex {

                }
            })
        }
    }

    @IBAction func forgotPasswordButtonPressed(_ sender: Any) {
        forgotPasswordVC = STForgotPasswordViewController(nibName: "STForgotPasswordViewController", bundle: Bundle.main)
        forgotPasswordVC?.delegate = self
        forgotPasswordVC?.view?.frame = view.frame
        view.addSubview(forgotPasswordVC?.view, withAnimation: true)
    }

    @IBAction func logIn(viaFacebookButtonPressed sender: Any) {
        STLogInManager.sharedInstance().delegate = self
        STLogInManager.sharedInstance().openSession(withAllowLoginUI: self)
    }

    func sendPaswordToEmailButtonPressed() {
        if forgotPasswordVC != nil {
            forgotPasswordVC?.view?.remove(fromSuperviewwithAnimation: true)
        }
    }

    func forgetpasswordCanceled() {
        if forgotPasswordVC != nil {
            forgotPasswordVC?.view?.remove(fromSuperviewwithAnimation: true)
        }
    }

    @IBAction func backButtonPressed(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }

    func rightGestureHandle() {
        navigationController?.popViewController(animated: true)
        STSharedCustoms.sharedAddGestureInstance(withDelegate: nil)
    }
// MARK: ----- Custom Methods

    func launchUserBaseTabsScreen() {
        var storyboard = UIStoryboard(name: "Main", bundle: Bundle.main)
        var baseTabBarController: STBaseTabBarController? = storyboard.instantiateViewController(withIdentifier: "STBaseTabBarController")
        navigationController?.pushViewController(baseTabBarController, animated: true)
    }
// MARK: ----- Textfield Methods

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        if textField == txtFieldUsernameId {
            txtFieldPassword.becomeFirstResponder()
        }
        else {
            textField.resignFirstResponder()
        }
        return true
    }

    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        var shouldChange: Bool = true
        if textField == txtFieldPassword {
            if string.validateAlphanumeric() {
                shouldChange = true
            }
            else {
                shouldChange = false
            }
        }
        return shouldChange
    }
// MARK: ----- LogInManagerDelegate

    func facebookLogInSuccessful(withUserDictionary userInfoDict: [AnyHashable: Any]) {
        print("facebooklogIn Dict \(userInfoDict)")
        if userInfoDict != nil {
            var localUserInfoDict = [AnyHashable: Any]()
            localUserInfoDict[kParamKeyFirstname] = userInfoDict["first_name"]
            localUserInfoDict[kParamKeyLastname] = userInfoDict["last_name"]
            localUserInfoDict[kParamKeyGender] = userInfoDict["gender"]
            localUserInfoDict[kParamKeyEmail] = userInfoDict["email"]
                //[localUserInfoDict setObject:[[userInfoDict objectForKey:@"location"] objectForKey:@"name"] forKey:kParamKeyCountry];
            var dateFormat = DateFormatter()
            dateFormat.dateFormat = "MM/dd/yyyy"
            var theDate: Date? = dateFormat.date(fromString: (userInfoDict["birthday"] as? Date))
            var dateFormat2 = DateFormatter()
            dateFormat2.dateFormat = "yyyy-MM-dd"
            if dateFormat2.string(from: theDate) {
                localUserInfoDict[kParamKeyDateOfBirth] = dateFormat2.string(from: theDate)
            }
            if Data(contentsOf: URL(string: "http://graph.facebook.com/\(userInfoDict["id"])/picture")).base64EncodedString(withOptions: 0) != nil {
                localUserInfoDict[kParamKeyAvatar] = Data(contentsOf: URL(string: "http://graph.facebook.com/\(userInfoDict["id"])/picture")).base64EncodedString(withOptions: 0)
            }
            if httpReq != nil {
                httpReq?.delegate = nil
                httpReq = nil
            }
            httpReq = BOABHttpReq.initBoabReq(withDelegate: self, shouldShowActivityIndicatorView: true)
            localUserInfoDict[kParamKeyAction] = kAPIActionFacebookLogIn
            httpReq?.sendAsyncPostRequest(STASHER_BASE_URL, params: localUserInfoDict, json: true)
        }
    }
// MARK: - BOABHttpReqDelegates

    func boabHttpReqFinished(withResponse resonseData: Data, andConnection conn: BOABURLConnection) {
        view.userInteractionEnabled = true
        var actionName: String? = ((conn.userInfo?[KParamsKey] as? String)?[kParamKeyAction] as? String)
        if (actionName == kAPIActionFacebookLogIn) {
            UserDefaults.standard.set("YES", forKey: "IsFaceBookLogIn")
        }
        else {
            UserDefaults.standard.set("NO", forKey: "IsFaceBookLogIn")
        }
        UserDefaults.standard.synchronize()
        if resonseData {
            var responseDictionary: [AnyHashable: Any]? = try? JSONSerialization.jsonObject(withData: resonseData, options: kNilOptions)
            if responseDictionary != nil {
                if responseDictionary?["usedetails"] && (responseDictionary?["usedetails"] is [AnyHashable: Any]) {
                    STUserIdentity.sharedInstance().userInformationDictionary = responseDictionary?["usedetails"]
                    var logInInfoDictionary: [AnyHashable: Any]? = (responseDictionary?["usedetails"] as? [AnyHashable: Any])?
                    logInInfoDictionary?[kUserDefaultsIsUserLoggedIn] = Int(1)
                    STLogInManager.sharedInstance().updateUserDefaultsInLogInManager(withDictionary: logInInfoDictionary)
                    launchUserBaseTabsScreen()
                }
                else if responseDictionary?["error"] {
                    UIAlertView.show(withTitle: "", message: responseDictionary?["error"]["message"], cancelButtonTitle: "OK", otherButtonTitles: nil, tapBlock: {(_ alertView: UIAlertView, _ buttonIndex: Int) -> Void in
                        if buttonIndex == alertView.cancelButtonIndex {

                        }
                    })
                }
            }
        }
    }

    func boabHttpReqFailedWithError(_ error: Error?) {
        view.userInteractionEnabled = true
        if error != nil {

        }
    }
}